java/ql/src/experimental/Security/CWE/CWE-600/UncaughtServletException.ql

```diff
@@ -0,0 +1,85 @@
+/**
+ * @name Uncaught Servlet Exception
+ * @description Uncaught exceptions in a servlet could leave a system in a vulnerable state, possibly resulting in denial-of-service attacks or the exposure of sensitive information disclosed in stack traces.
+ * @kind path-problem
+ * @id java/uncaught-servlet-exception
+ * @tags security
+ *       external/cwe-600
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.frameworks.Servlets
+import semmle.code.xml.WebXML
+import DataFlow::PathGraph
+
+/** The type `java.io.IOException`. */
+class IOException extends RefType {
+  IOException() { this.hasQualifiedName("java.io", "IOException") }
+}
+
+/** Check whether a given exception type is caught. */
+private predicate catchesEx(TryStmt t, RefType exType) {
+  exists(CatchClause cc, LocalVariableDeclExpr v |
+    t.getACatchClause() = cc and
+    cc.getVariable() = v and
+    v.getType().(RefType).getASubtype*() = exType //Detect the case that a subclass exception is thrown but its parent class is declared in the catch clause.
+  )
+}
+
+/** Servlet methods of `javax.servlet.http.Servlet` and subtypes. */
+private predicate isServletMethod(Callable c) {
+  c.getDeclaringType() instanceof ServletClass and
+  c.getNumberOfParameters() = 2 and
+  c.getParameter(1).getType() instanceof ServletResponse and
+  (
+    c.getName() = "doGet" or
+    c.getName() = "doPost" or
+    c.getName() = "doPut" or
+    c.getName() = "doDelete" or
+    c.getName() = "doHead" or
+    c.getName() = "doOptions" or
+    c.getName() = "doTrace" or
+    c.getName() = "service"
+  )
+}
+
+/** Holds if `web.xml` has an error page configured. */
+private predicate hasErrorPage() {
+  exists(WebErrorPage wep | wep.getPageLocation().getValue() != "")
+}
+
+/** Sink of uncaught IO exceptions or runtime exceptions since other exception types must be explicitly caught. */
+class UncaughtServletExceptionSink extends DataFlow::ExprNode {
+  UncaughtServletExceptionSink() {
+    exists(Method m, MethodAccess ma | ma.getMethod() = m |
+      isServletMethod(ma.getEnclosingCallable()) and
+      (
+        m.getAThrownExceptionType().getASupertype*() instanceof IOException or
+        m
+            .getAThrownExceptionType()
+            .getASupertype*()
+            .hasQualifiedName("java.lang", "RuntimeException")
```

Comment:
Why restrict to particular exception types? Doesn't it suffice to check if a particular exception is thrown and not caught within this method, since this is the top-level doGet or similar method?

Reply:
As I mentioned in the comments, uncaught exceptions shall be IO exceptions or runtime exceptions since other exception types must be explicitly caught. I think the query behaviours the same without explicitly specifying those exception types. I've removed from the query.